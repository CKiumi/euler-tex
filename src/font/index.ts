import { INTER_ATOM_SPACING, SIGMAS, AtomKind } from "./sigma";
import { Font, SPEC } from "./spec";
import METRICS from "./data";
export { SPEC, SIGMAS };
export type { Font, AtomKind };

export type CharMetric = {
  depth: number;
  height: number;
  italic: number;
  skew: number;
  width: number;
};

export const getCharMetrics = (char: string, font: Font): CharMetric => {
  try {
    const metrics = METRICS[font][char.charCodeAt(0)];
    return {
      depth: metrics[0],
      height: metrics[1],
      italic: metrics[2],
      skew: metrics[3],
      width: metrics[4],
    };
  } catch (error) {
    throw new Error(`Font metrics not found for font: ${font}.`);
  }
};

export const getSigma = (name: keyof typeof SIGMAS): number => {
  return SIGMAS[name][0];
};

export const getSpacing = (prevKind: AtomKind, curKind: AtomKind): number =>
  (INTER_ATOM_SPACING[prevKind]?.[curKind] ?? 0) * CSSEmPerMu;

const CSSEmPerMu = SIGMAS.quad[0] / 18;
