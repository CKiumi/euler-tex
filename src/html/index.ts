export * from "./svg/inner";
export * from "./svg/pathNode";
export * from "./svg/sqrt";

export const html = <K extends keyof HTMLElementTagNameMap>(
  tagName: K,
  options: {
    text?: string;
    cls?: string[];
    id?: string;
    style?: Partial<CSSStyleDeclaration>;
    children?: HTMLElement[];
  }
): HTMLElementTagNameMap[K] => {
  const span = document.createElement(tagName);
  Object.assign(span.style, options.style);
  options.id && (span.id = options.id);
  options.cls && span.classList.add(...options.cls);
  options.text && (span.innerText = options.text);
  options.children && span.append(...options.children);
  return span;
};
