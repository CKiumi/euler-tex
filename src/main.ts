import "./style.css";
import katex from "katex";
import "katex/dist/katex.min.css";
const app = document.querySelector<HTMLDivElement>("#app");
if (app) {
  app.innerHTML = "<h1>Euler Tex</h1>";
  katex.render("c = \\pm\\sqrt{a^2 + b^2}", app, {
    throwOnError: false,
  });
}
