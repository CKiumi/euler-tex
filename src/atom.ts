import { Box, HBox, SymBox, toVBox, VStackBox } from "/src/box";
import { getCharMetrics, getSigma, getSpacing } from "/font";
import { AtomKind } from "/font/src/sigma";
import { Font } from "/font/src/spec";

export interface Atom {
  type: "sym" | "accent" | "overline" | "line";
  kind: AtomKind;
}

export interface SymAtom extends Atom {
  type: "sym";
  char: string;
  font: Font;
}

export interface AccentAtom extends Atom {
  type: "accent";
  body: Atom;
  accent: Accent;
}

export interface Accent extends SymAtom {
  char: "^" | "~";
  font: "Main-R";
  kind: "ord";
}

export interface OverlineAtom extends Atom {
  type: "overline";
  body: Atom;
}
export interface LineAtom extends Atom {
  type: "line";
}

export const parseAtoms = (atoms: Atom[]): HBox => {
  let prevKind: AtomKind | undefined;
  const children = atoms.map((atom) => {
    const box = parseAtom(atom);
    if (prevKind && atom.kind) {
      box.spacing = getSpacing(prevKind, atom.kind);
    }
    prevKind = atom.kind;
    return box;
  });
  const width = children.reduce((acc, a) => acc + a.width, 0);
  const depth = Math.max(...children.map((child) => child.depth));
  const height = Math.max(...children.map((child) => child.height));
  return { children, width, height, depth };
};

export const parseAtom = (atom: Atom): Box => {
  if (atom.type === "sym") return parseSymAtom(atom as SymAtom);
  if (atom.type === "overline") return parseOverline(atom as OverlineAtom);
  if (atom.type === "line") return parseLine();
  if (atom.type === "accent") return parseAccentAtom(atom as AccentAtom);
  throw new Error("No Atom Type Specified");
};

export const parseSymAtom = ({ char, font }: SymAtom): SymBox => {
  const { depth, height, italic, width } = getCharMetrics(char, font);
  return { char, font, depth, height, width: width + italic, italic };
};

export const parseAccentAtom = (atom: AccentAtom): VStackBox => {
  const box = parseSymAtom(atom.body as SymAtom);
  const accBox = parseSymAtom(atom.accent);
  const clearance = Math.min(box.height, getSigma("xHeight"));
  accBox.spacingBelow = -clearance;
  return toVBox([accBox, box], box.depth);
};

export const parseOverline = (body: OverlineAtom): VStackBox => {
  const accBox = parseAtom({ type: "line", kind: "ord" });
  const box = parseSymAtom(body.body as SymAtom);
  const defaultRuleThickness = getSigma("defaultRuleThickness");
  accBox.spacingTop = defaultRuleThickness;
  accBox.spacingBelow = 3 * defaultRuleThickness;
  return toVBox([accBox, box], box.depth);
};

export const parseLine = (): Box => {
  return { width: 0, height: getSigma("defaultRuleThickness"), depth: 0 };
};
