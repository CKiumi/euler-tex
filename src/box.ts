import { Atom, SymAtom } from "./atom";
import { getCharMetrics } from "/font";
import { Font } from "/font/src/spec";

export interface Box {
  height: number;
  depth: number;
  width: number;
}

export interface SymBox extends Box {
  char: string;
  font: Font;
  height: number;
  depth: number;
  width: number;
  italic: number;
}

export interface HBox extends Box {
  children: Box[];
  height: number;
  depth: number;
  width: number;
}

export const toHBox = (atoms: Atom[]): HBox => {
  const children = atoms.map(toSymBox);
  const width = children.reduce((acc, a) => acc + a.width, 0);
  const depth = Math.max(...children.map((child) => child.depth));
  const height = Math.max(...children.map((child) => child.height));
  return { children, width, height, depth };
};

export const toSymBox = ({ char, font }: SymAtom): SymBox => {
  const { depth, height, italic, width } = getCharMetrics(char, font);
  return { char, font, depth, height, width: width + italic, italic };
};
