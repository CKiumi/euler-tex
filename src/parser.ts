import { Atom, LRAtom, SqrtAtom, SymAtom } from "./atom/atom";
import { SupSubAtom } from "./atom/supsub";
import { Escape, Lexer, Token } from "./lexer";
import { AtomKind } from "/font/src/sigma";

export class Parser {
  lexer: Lexer;
  constructor(lexer: Lexer) {
    this.lexer = lexer;
  }
  parseSingle(atoms: Atom[], token: Token) {
    if (/[A-Za-z]/.test(token)) return new SymAtom("ord", token, "Math-I");
    if (/[0-9]/.test(token)) return new SymAtom("ord", token, "Main-R");
    if (token === "+") return new SymAtom(binOrOrd(atoms), "+", "Main-R");
    if (token === "-") return new SymAtom(binOrOrd(atoms), "−", "Main-R");
    if (token === "=") return new SymAtom("rel", "=", "Main-R");
    if (token === ",") return new SymAtom("punct", ",", "Main-R");
    throw new Error(`Single token ${token} not supported`);
  }
  parse(end: Token): Atom[] {
    const atoms: Atom[] = [];
    for (;;) {
      const token = this.lexer.tokenize();
      if (token === end) break;
      if (token === Escape.EOF) throw new Error(`Expected ${end}`);
      if (token.length === 1) atoms.push(this.parseSingle(atoms, token));
      if (token === Escape.Left) atoms.push(this.parseLR());
      if (token === Escape.Circumfix) {
        const body = atoms.pop();
        if (!body) throw new Error("body must exist");
        if (body instanceof SupSubAtom) {
          if (!body.sub) throw new Error("sub must exist");
          atoms.push(new SupSubAtom(body.nuc, this.parseArg(atoms), body.sub));
        } else
          atoms.push(new SupSubAtom(body, this.parseArg(atoms), undefined));
      }
      if (token === Escape.UnderScore) {
        const body = atoms.pop();
        if (!body) throw new Error("body must exist");
        if (body instanceof SupSubAtom) {
          if (!body.sup) throw new Error("sup must exist");
          atoms.push(new SupSubAtom(body.nuc, body.sup, this.parseArg(atoms)));
        } else
          atoms.push(new SupSubAtom(body, undefined, this.parseArg(atoms)));
      }
      if (token.startsWith("\\")) {
        if (token === "\\sum") atoms.push(new SymAtom("op", "∑", "Size2"));
        if (token === "\\int") atoms.push(new SymAtom("op", "∫", "Size2"));
        if (token === "\\sqrt") atoms.push(new SqrtAtom(this.parseArg(atoms)));
        const key = Object.keys(LETTER).find((l) => token === l);
        if (key) atoms.push(new SymAtom("ord", LETTER[key], "Math-I"));
      }
    }
    return atoms;
  }
  private parseLR(): Atom {
    const left = this.lexer.tokenize();
    const body = this.parse(Escape.Right);
    this.lexer.tokenize();
    return new LRAtom(left, ")", body) as Atom;
  }
  parseArg(atoms: Atom[]) {
    const token = this.lexer.tokenize();
    if (token === Escape.LCurly) return this.parse(Escape.RCurly);
    else return [this.parseSingle(atoms, token)];
  }
}

const binOrOrd = (atoms: Atom[]): AtomKind => {
  const kind = /bin|op|rel|open|punct/.test(
    atoms[atoms.length - 1].kind ?? "bin"
  )
    ? "ord"
    : "bin";
  return kind;
};

export const parse = (latex: string): Atom[] => {
  return new Parser(new Lexer(latex)).parse(Escape.EOF);
};

export const LETTER: { [key: string]: string } = {
  "\\alpha": "α",
  "\\beta": "β",
  "\\gamma": "γ",
  "\\delta": "δ",
  "\\epsilon": "ε",
  "\\lambda": "λ",
  "\\zeta": "ζ",
  "\\eta": "η",
};
