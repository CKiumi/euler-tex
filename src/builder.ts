import { HBox, SymBox } from "./box";
import { em } from "./util";
import { SPEC } from "/font/src/spec";

export const buildHBox = (hbox: HBox) => {
  const span = document.createElement("span");
  span.classList.add("hbox");
  hbox.children.forEach((box) => {
    span.append(buildSymBox(box as SymBox));
  });
  return span;
};

export const buildSymBox = ({
  char,
  font,
  height,
  depth,
  italic,
  spacing,
}: SymBox): HTMLSpanElement => {
  const span = document.createElement("span");
  span.innerText = char;
  span.classList.add("box", font.toLowerCase());
  span.style.height = em(height + depth);
  if (italic) {
    span.style.paddingRight = em(italic);
  }
  if (spacing) {
    span.style.marginLeft = em(spacing);
  }
  span.style.lineHeight = em(
    (height + (SPEC[font].descent - SPEC[font].ascent) / 2) * 2
  );
  return span;
};
