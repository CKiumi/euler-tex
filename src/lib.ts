import { GroupAtom } from "./atom/atom";
import { parse } from "./parser/parser";
export * from "./font";
export * from "./atom/atom";
export * from "./box/box";
export * from "./parser/parser";

export const latexToHtml = (latex: string) =>
  new GroupAtom(parse(latex)).toBox().toHtml();

export const latexToEditableAtom = (latex: string) => {
  const atom = new GroupAtom(parse(latex, true), true);
  atom.toBox().toHtml();
  return atom;
};
