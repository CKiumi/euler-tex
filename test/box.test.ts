import { tests } from "./../src/main";
import { expect, test } from "vitest";

test("first", () => {
  expect(tests()).matchSnapshot();
});
