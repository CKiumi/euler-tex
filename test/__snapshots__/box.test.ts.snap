// Vitest Snapshot v1

exports[`first 1`] = `<span />`;
